Handle form choice duplicates

Hlavtox opened the pull request:
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | 9.0.x
| Description?      | Introduces a class that formats choices for symfony forms and automatically marks duplicate options with an ID, so we avoid having choices that just disappear. Initally implemented on Manufacturer, Supplier, Feature value and Carrier choices, but easy to do on the rest.
| Type?             | refacto
| Category?         | BO
| BC breaks?        | yes
| Deprecations?     | no
| How to test?      | Create two manufacturers with the same name, go to product form, see that you can see both.
| UI Tests          | https://github.com/florine2623/testing_pr/actions/runs/11288952311 ✅ 
| Fixed issue or discussion?     | Fixes https://github.com/PrestaShop/PrestaShop/issues/36586
| Related PRs       | 
| Sponsor company   |

Hlavtox committed:
Add FormChoiceFormatter

Hlavtox commented:
Ping @matks @jolelievre @kpodemski to check the concept, before I move on. Is it OK?

Hlavtox commented:
@PrestaShop/committers Hi, can someone review please? :-)

FabienPapet commented on src/PrestaShopBundle/Resources/config/services/adapter/form.yml:
```diff
@@ -79,12 +80,15 @@ services:
   prestashop.adapter.form.choice_provider.supplier_name_by_id_choice_provider:
     class: 'PrestaShop\PrestaShop\Adapter\Form\ChoiceProvider\SupplierNameByIdChoiceProvider'
+    arguments:
+      - '@PrestaShop\PrestaShop\Core\Util\FormChoiceFormatter'
```
```suggestion
    autowire: true
```

FabienPapet commented on src/PrestaShopBundle/Resources/config/services/adapter/data_provider.yml:
```diff
@@ -29,7 +29,9 @@ services:
   prestashop.adapter.data_provider.carrier:
     class: PrestaShop\PrestaShop\Adapter\Carrier\CarrierDataProvider
-    arguments: [ '@prestashop.adapter.legacy.configuration' ]
+    arguments:
+      - '@prestashop.adapter.legacy.configuration'
+      - '@PrestaShop\PrestaShop\Core\Util\FormChoiceFormatter'
```
```suggestion
    autowire: true
```

FabienPapet commented on src/Adapter/Carrier/CarrierDataProvider.php:
```diff
@@ -39,9 +40,17 @@ class CarrierDataProvider
+    /**
+     * @var FormChoiceFormatter
+     */
+    private $formChoiceFormatter;
```
```suggestion
    private FormChoiceFormatter $formChoiceFormatter;
```

FabienPapet commented:
I'd prefer having a static class instead of using DI in this specific case but it is not blocking 😀

Hlavtox commented:
@FabienPapet Changed to static class and added a test. ;-)

FabienPapet commented:
I love your test data

FabienPapet commented on tests/Unit/Core/Util/FormChoiceFormatterTest.php:
```diff
@@ -0,0 +1,109 @@
+     * @param string $nameKey
+     * @param array $expectedFormattedChoices
+     */
+    public function testCleanUrl(array $rawOptions, string $idKey, string $nameKey, array $expectedFormattedChoices): void
```
Copy pasted ;) not blocking

Hlavtox replied:
Ha, good catch, I will fix it. Also copied `clean absolute url _token`.

Hlavtox commented:
Ping @jolelievre @matks @boherm

florine2623 commented:
@Hlavtox , UI tests are red, could you check that ? 
https://github.com/florine2623/testing_pr/actions/runs/11105865072

@PrestaShop/qa-automation , what do you think about this ? 

Thanks to all of you !

Progi1984 commented:
@florine2623 The nightly is :green_circle: so it's related to the PR.

![image](https://github.com/user-attachments/assets/ce0e0f9a-8d5f-4ecf-afbf-c27311ddaed3)

jolelievre commented on src/Core/Util/FormChoiceFormatter.php:
```diff
@@ -0,0 +1,89 @@
+
+declare(strict_types=1);
+
+namespace PrestaShop\PrestaShop\Core\Util;
```
```suggestion
namespace PrestaShop\PrestaShop\Core\Form;
```

jolelievre replied:
Form namespace seems more accurate for this tool

jolelievre commented on src/Core/Util/FormChoiceFormatter.php:
```diff
@@ -0,0 +1,89 @@
+        $finalChoices = [];
+
+        // A slim array with just they keys we processed, so we know what are duplicates
+        $alreadyProcessedKeys = [];
```
```suggestion
        $alreadyProcessedNames = [];
```

jolelievre commented on src/Core/Util/FormChoiceFormatter.php:
```diff
@@ -0,0 +1,89 @@
+            $alreadyProcessedKeys[] = $rawChoice[$nameKey];
+
+            // And save some memory, we don't need it anymore
+            unset($rawChoices[$rawChoiceKey]);
```
Not sure this is needed or if it brings any improvement on performances 🤔 You can simply loop through the array without modifying it If so you can also simplify the for loop:
```php
foreach ($rawChoices as $rawChoice) {
```

Hlavtox replied:
My reasoning was that I gradually cut from one array and add it, so we don't increase memory usage by looping it. :-)

jolelievre commented on tests/Unit/Core/Util/FormChoiceFormatterTest.php:
```diff
@@ -0,0 +1,109 @@
+
+declare(strict_types=1);
+
+namespace Tests\Unit\Core\Util;
```
```suggestion
namespace Tests\Unit\Core\Form;
```

jolelievre commented on tests/Unit/Core/Util/FormChoiceFormatterTest.php:
```diff
@@ -0,0 +1,109 @@
+     * @param string $nameKey
+     * @param array $expectedFormattedChoices
+     */
+    public function testCleanUrl(array $rawOptions, string $idKey, string $nameKey, array $expectedFormattedChoices): void
```
```suggestion
    public function testFormatFormChoices(array $rawOptions, string $idKey, string $nameKey, array $expectedFormattedChoices): void
```

jolelievre commented on tests/Unit/Core/Util/FormChoiceFormatterTest.php:
```diff
@@ -0,0 +1,109 @@
+
+    public function getFormOptionsToFormat(): iterable
+    {
+        yield 'clean absolute url _token' => [
```
```suggestion
        yield 'manufacturer list with duplicates' => [
```

jolelievre commented on tests/Unit/Core/Util/FormChoiceFormatterTest.php:
```diff
@@ -0,0 +1,109 @@
+                'Preston Manufacturing' => 2,
+                'Trendo' => 3,
+            ],
+        ];
```
You should also add an use case where there is no duplicates, to validate that in this case the names are not modified at all (which is 90% of the time probably)

jolelievre commented on src/Core/Util/FormChoiceFormatter.php:
```diff
@@ -0,0 +1,89 @@
+        }
+
+        // Order data by displayed value (array key) by default
+        ksort($finalChoices);
```
Maybe, the sorting should be optional? With a `$sortByNames = true` by default, but in some cases the order can be made based on other criteria of importance and we shouldn't change it

If you add this option then a third use case with this option at `false` would be nice in the unit tests

jolelievre replied:
Oh actually what I thought as a suggestion for a potential use case is already required, the Carrier list has a specific order by position So in this case you need to disable the automatic ksort It is likely what makes the UI tests fail in this case

jolelievre replied:
The other failing test maybe be related to `ProfileByIdChoiceProvider` that also forces an order by ID internally, in this case I don't know if we should keep the original order or adapt the UI tests (but for carrier the order by position makes sense and should be prioritized)

jolelievre reviewed (commented):
Thanks @Hlavtox 

That's a great idea! Just a few suggestions, but I love the idea!

Hlavtox commented:
@jolelievre Addressed all feedback! :-)

- Added option to not use sorting.
- For this reason, I changed the way I replace previous values in the array, I now change the key keeping the position.
- Changed namespaces.
- Changed forgotten copy-pastes in tests.
- Added new tests, without duplicates and with duplicates and no sorting.

Let's see what the tests say!

PululuK commented:
Bien vu @Hlavtox 👍🏿

Hlavtox commented:
I removed some usages to not have to deal with tests now, they need to be adapter. It can be done in a second step where we implement this everywhere.

New test run - https://github.com/Hlavtox/ga.tests.ui.pr/actions/runs/11141582905

jolelievre commented on tests/Unit/Core/Form/FormChoiceFormatterTest.php:
```diff
@@ -0,0 +1,209 @@
+            ],
+        ];
+
+        yield 'manufacturer list with duplicates and disabled sorting' => [
```
Nide one! 👍

jolelievre reviewed (dismissed):
Thanks @Hlavtox

PululuK commented on src/Core/Form/FormChoiceFormatter.php:
```diff
@@ -0,0 +1,110 @@
+     *
+     * @return array Formatted choices
+     */
+    public static function formatFormChoices(array $rawChoices, string $idKey, string $nameKey, bool $sortByName = true): array
```
I think the method name is a bit redundant with the class name. `FormChoiceFormatter::format` would be sufficient

florine2623 commented:
Hello @Hlavtox and @PrestaShop/qa-automation 

I have relaunched the auto tests : https://github.com/florine2623/testing_pr/actions/runs/11288952311
They are still red. What should be done from there ?

Thanks!

florine2623 commented:
@Hlavtox , seen with @nesrineabdmouleh , the problem of the auto test comes from your PR.

Could you check it please ? :)
Thanks!

nesrineabdmouleh committed:
Fix UI tests

florine2623 commented:
Hello @Hlavtox ,

https://github.com/florine2623/testing_pr/actions/runs/11288952311 tests are green, thanks @nesrineabdmouleh !

It is QA ✅

Progi1984 reviewed (dismissed):
Ok for ui tests

nicosomb commented:
PHP CS fixer is not happy.

kpodemski reviewed (dismissed):
Hello

We recently introduced the 9.0.x branch, which will be the base for PrestaShop 9.0. If you want to see your change in this version, please rebase your PR onto 9.0.x. The `develop` will be used for PrestaShop 9.1, so you need to decide what to do.

Please rebase onto `9.0.x` or let me know your decision if you don't need to have it there so I can unblock the pull request.

Hlavtox committed:
Fix CS

Hlavtox commented:
@kpodemski Ready, code style fixed. Can we merge?

kpodemski commented:
@Hlavtox CI, second approval and I think we can merge because it had QA ✅ before the rebase

jolelievre reviewed (approved):
Thanks @Hlavtox
